File: jas_ocaml/lib/interpreter/length.ml
(* Unit-aware length parser / formatter — see length.mli for spec. *)

let supported_units = ["pt"; "px"; "in"; "mm"; "cm"; "pc"]

let pt_per_unit unit =
  match String.lowercase_ascii unit with
  | "pt" -> Some 1.0
  (* CSS reference 96 dpi: 1 px = 1/96 in, 1 pt = 1/72 in
     ⇒ 1 px = 72/96 = 0.75 pt. *)
  | "px" -> Some 0.75
  | "in" -> Some 72.0
  | "mm" -> Some (72.0 /. 25.4)
  | "cm" -> Some (720.0 /. 25.4)
  | "pc" -> Some 12.0
  | _ -> None

let is_ws c = c = ' ' || c = '\t' || c = '\n' || c = '\r'

(* String.trim — but the stdlib version is sufficient. Keep alias for
   readability at call sites. *)
let strip s = String.trim s

let parse input ~default_unit =
  let s = strip input in
  let n = String.length s in
  if n = 0 then None
  else begin
    let i = ref 0 in
    let num_start = !i in
    (* Optional leading sign. *)
    if !i < n && (s.[!i] = '-' || s.[!i] = '+') then incr i;
    let saw_digit_before_dot = ref false in
    let saw_dot = ref false in
    let saw_digit_after_dot = ref false in
    let scanning = ref true in
    while !scanning && !i < n do
      let c = s.[!i] in
      if c >= '0' && c <= '9' then begin
        if !saw_dot then saw_digit_after_dot := true
        else saw_digit_before_dot := true;
        incr i
      end else if c = '.' && not !saw_dot then begin
        saw_dot := true;
        incr i
      end else
        scanning := false
    done;
    (* Reject lone -, ., -. (no digits at all). *)
    if not !saw_digit_before_dot && not !saw_digit_after_dot then None
    else begin
      let num_str = String.sub s num_start (!i - num_start) in
      match float_of_string_opt num_str with
      | None -> None
      | Some value ->
        (* Skip whitespace between number and unit. *)
        while !i < n && is_ws s.[!i] do incr i done;
        let unit_start = !i in
        let is_alpha c =
          (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
        in
        while !i < n && is_alpha s.[!i] do incr i done;
        let unit_end = !i in
        (* Trailing whitespace permitted; anything else is garbage. *)
        let rest_ws = ref true in
        while !rest_ws && !i < n do
          if is_ws s.[!i] then incr i else rest_ws := false
        done;
        if !i < n then None
        else begin
          let unit_str =
            if unit_end > unit_start then
              String.sub s unit_start (unit_end - unit_start)
            else default_unit in
          match pt_per_unit unit_str with
          | None -> None
          | Some factor -> Some (value *. factor)
        end
    end
  end

(* Trim trailing zeros and a stranded trailing decimal point from a
   formatted decimal. "12.50" -> "12.5"; "12.00" -> "12"; "12" -> "12". *)
let trim_zeros s =
  if not (String.contains s '.') then s
  else begin
    let n = String.length s in
    let i = ref (n - 1) in
    while !i > 0 && s.[!i] = '0' do decr i done;
    if s.[!i] = '.' then decr i;
    String.sub s 0 (!i + 1)
  end

let format pt ~unit ~precision =
  match pt with
  | None -> ""
  | Some pt when not (Float.is_finite pt) -> ""
  | Some pt ->
    let display_unit, factor =
      match pt_per_unit unit with
      | Some f -> String.lowercase_ascii unit, f
      | None -> "pt", 1.0 in
    let value = pt /. factor in
    (* `+. 0.0` normalises -0.0 → 0.0 before formatting. *)
    let formatted = Printf.sprintf "%.*f" precision (value +. 0.0) in
    let trimmed = trim_zeros formatted in
    let normalized = if trimmed = "-0" then "0" else trimmed in
    Printf.sprintf "%s %s" normalized display_unit

File: workspace_interpreter/length.py
"""Unit-aware length parser / formatter.

The single source of truth for the length-conversion table used by the
`length_input` widget. Flask, Python (jas), and any other Python
consumer share this; the JS port in jas_flask/static/js/app.js and the
Rust port in jas_dioxus/src/interpreter/length.rs follow the same
contract.

See UNIT_INPUTS.md for the spec these helpers implement.
"""

from __future__ import annotations

import re
from typing import Optional


SUPPORTED_UNITS = ("pt", "px", "in", "mm", "cm", "pc")
"""Units accepted by `parse_length` / produced by `format_length`."""


# Conversion: how many points are in one of the named unit. The
# canonical storage unit is pt — everywhere a length is committed to
# state or written to SVG, it's a pt number. Flask / Rust / Swift /
# OCaml / Python all keep their copies in sync with this table.
_PT_PER_UNIT = {
    "pt": 1.0,
    # CSS reference dpi convention: 1 px = 1/96 in, 1 pt = 1/72 in
    # ⇒ 1 px = 72/96 pt = 0.75 pt.
    "px": 0.75,
    "in": 72.0,
    # 1 in = 25.4 mm.
    "mm": 72.0 / 25.4,
    "cm": 720.0 / 25.4,
    # 1 pica = 12 pt.
    "pc": 12.0,
}


# Grammar (case-insensitive on the unit; case doesn't apply to the
# number): optional sign, then either `digits[.digits?]` or `.digits`,
# then optional whitespace, then optional unit, then trailing space.
# The regex is anchored so anything left over after the unit is a
# parse failure (rejecting "12pt5" / "12 mm 5" / etc.).
_LENGTH_RE = re.compile(
    r"""
    ^
    \s*
    (?P<num>-? (?: \d+ \.? \d* | \. \d+ ))    # number
    \s*
    (?P<unit>[A-Za-z]+)?                       # optional unit
    \s*
    $
    """,
    re.VERBOSE,
)


def pt_per_unit(unit: str) -> Optional[float]:
    """Return the pt-equivalent of one of the named unit, or None for
    unsupported names."""
    if not isinstance(unit, str):
        return None
    return _PT_PER_UNIT.get(unit.lower())


def parse_length(s: str, default_unit: str) -> Optional[float]:
    """Parse a user-typed length string into a value in points.

    Bare numbers are interpreted in `default_unit`. Unit suffixes
    override the default. Whitespace is tolerated around / between the
    number and the unit. Returns `None` when the input is empty,
    syntactically malformed, or carries an unsupported unit.

    Per UNIT_INPUTS.md §Edge cases — callers decide whether `None`
    means "commit null on a nullable field" or "revert".
    """
    if not isinstance(s, str):
        return None
    if not s.strip():
        return None

    m = _LENGTH_RE.match(s)
    if m is None:
        return None

    num_str = m.group("num")
    unit_str = m.group("unit")

    # The regex's number alternation accepts a lone `-` or `.`; reject
    # those explicitly — Python's float() would also fail, but the
    # explicit check is clearer at the call site.
    try:
        value = float(num_str)
    except ValueError:
        return None

    unit = (unit_str or default_unit).lower()
    factor = _PT_PER_UNIT.get(unit)
    if factor is None:
        return None
    return value * factor


def format_length(pt: Optional[float], unit: str, precision: int = 2) -> str:
    """Render a pt value as a display string in the named unit.

    `None` formats as an empty string (used by nullable dash / gap
    fields when no value is set). Trailing zeros and a trailing decimal
    point are trimmed. Unknown / unsupported `unit` falls back to pt
    rather than producing a malformed output.
    """
    if pt is None:
        return ""

    factor = _PT_PER_UNIT.get(unit.lower()) if isinstance(unit, str) else None
    if factor is None:
        unit = "pt"
        factor = 1.0

    value = pt / factor
    rounded = round(value, precision)
    # `round` may yield -0.0 on negative-near-zero inputs; normalise.
    if rounded == 0.0:
        rounded = 0.0
    text = f"{rounded:.{precision}f}"
    # Trim trailing zeros in the decimal part, then a stranded
    # trailing decimal point.
    if "." in text:
        text = text.rstrip("0").rstrip(".")
    return f"{text} {unit.lower()}"
